### PR TITLE
CBMC: Add spec + proof for gen_matrix_entry_x4 and gen_matrix

### LINKED ISSUE #322: CBMC State + prove spec for gen_matrix
Depends on: #326 #327

### LINKED ISSUE #327: CBMC: State and prove spec for gen_matrix_entry_x4

### PR DESCRIPTION
**Pull request by hanno-becker**
Much harder than anticipated. See the individual commit messages for detailed descriptions of the pre-work that had to happen.

- Resolves #327 
- Resolves #322

### CODE REVIEW

File: cbmc/proofs/gen_matrix_entry_x4/Makefile [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+INCLUDES +=
+
+REMOVE_FUNCTION_BODY +=
+UNWINDSET +=
+
+PROOF_SOURCES += $(PROOFDIR)/$(HARNESS_FILE).c
+PROJECT_SOURCES += $(SRCDIR)/mlkem/indcpa.c $(SRCDIR)/fips202/fips202x4.c
+
+CHECK_FUNCTION_CONTRACTS= gen_matrix_entry_x4 # rej_uniform_x4
+USE_FUNCTION_CONTRACTS=$(FIPS202_NAMESPACE)shake128x4_absorb $(FIPS202_NAMESPACE)shake128x4_squeezeblocks $(MLKEM_NAMESPACE)rej_uniform # rej_uniform_x4
```
Comment (mkannwischer) [inline cbmc/proofs/gen_matrix_entry_x4/Makefile:22]:
```suggestion
USE_FUNCTION_CONTRACTS=$(FIPS202_NAMESPACE)shake128x4_absorb $(FIPS202_NAMESPACE)shake128x4_squeezeblocks $(MLKEM_NAMESPACE)rej_uniform
```

File: cbmc/proofs/gen_matrix_entry_x4/Makefile [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+DEFINES +=
+INCLUDES +=
+
+REMOVE_FUNCTION_BODY +=
+UNWINDSET +=
+
+PROOF_SOURCES += $(PROOFDIR)/$(HARNESS_FILE).c
+PROJECT_SOURCES += $(SRCDIR)/mlkem/indcpa.c $(SRCDIR)/fips202/fips202x4.c
+
+CHECK_FUNCTION_CONTRACTS= gen_matrix_entry_x4 # rej_uniform_x4
```
Comment (mkannwischer) [inline cbmc/proofs/gen_matrix_entry_x4/Makefile:21]:
```suggestion
CHECK_FUNCTION_CONTRACTS= gen_matrix_entry_x4
```

File: mlkem/indcpa.c [hunk 3]
```diff
@@ -237,7 +266,15 @@ void gen_matrix(polyvec *a, const uint8_t seed[MLKEM_SYMBYTES],
                 int transposed) {
   int i;
   // We need MLKEM_SYMBYTES + 2 bytes per seed, but add padding for alignment
```
Comment (mkannwischer) [inline mlkem/indcpa.c:268]:
```suggestion
```

File: mlkem/indcpa.c [hunk 4]
```diff
@@ -147,10 +147,28 @@ static void unpack_ciphertext(polyvec *b, poly *v,
...
+{
   // Temporary buffers for XOF output before rejection sampling
-  uint8_t bufx[KECCAK_WAY][GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
+  uint8_t buf0[GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
+  uint8_t buf1[GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
+  uint8_t buf2[GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
+  uint8_t buf3[GEN_MATRIX_NBLOCKS * SHAKE128_RATE];
+
   // Tracks the number of coefficients we have already sampled
   unsigned int ctr[KECCAK_WAY];
```
Comment (mkannwischer) [inline mlkem/indcpa.c:173]: should this be changed to int?

**Review by mkannwischer — CHANGES_REQUESTED** [review]
Thanks the loop invariants and bounds look good to me. Just a few cosmetic changes please.

Please remove the comment: 
```// TODO: All loops in this function should be unrolled for decent
  // performance.```

File: mlkem/indcpa.c (hunk 3 above)
Comment (hanno-becker) [reply mlkem/indcpa.c]: Done

File: cbmc/proofs/gen_matrix/gen_matrix_harness.c [hunk 5]
```diff
@@ -0,0 +1,30 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+/*
+ * Insert copyright notice
+ */
+
+/**
+ * @file rej_uniform_harness.c
+ * @brief Implements the proof harness for rej_uniform function.
```
Comment (mkannwischer) [inline cbmc/proofs/gen_matrix/gen_matrix_harness.c:10]: please update.

File: cbmc/proofs/gen_matrix_entry_x4/gen_matrix_entry_x4_harness.c [hunk 6]
```diff
@@ -0,0 +1,32 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT-0
+
+/*
+ * Insert copyright notice
+ */
+
+/**
+ * @file rej_uniform_harness.c
+ * @brief Implements the proof harness for rej_uniform function.
```
Comment (mkannwischer) [inline cbmc/proofs/gen_matrix_entry_x4/gen_matrix_entry_x4_harness.c:10]: same.

**Review by mkannwischer — CHANGES_REQUESTED** [review]

File: cbmc/proofs/gen_matrix_entry_x4/gen_matrix_entry_x4_harness.c (hunk 6 above)
Comment (hanno-becker) [reply cbmc/proofs/gen_matrix_entry_x4/gen_matrix_entry_x4_harness.c]: done

File: cbmc/proofs/gen_matrix/gen_matrix_harness.c (hunk 5 above)
Comment (hanno-becker) [reply cbmc/proofs/gen_matrix/gen_matrix_harness.c]: done

**Review by mkannwischer — APPROVED** [review]

**Pull request merged**